File: 0733-flood-fill/0733-flood-fill.cpp
class Solution
{
    public:
       	//  in this case we don't need a separate visited structure: changing the color(2) guarantees we will not go to the same cell again.
        vector<vector < int>> floodFill(vector<vector < int>> &image, int sr, int sc, int color)
        {
            int n = image.size();
            int m = image[0].size();

            queue<pair<int, int>> q;

            q.push({ sr,
                sc });

            int startColor = image[sr][sc];

            int dx[] = { -1,
                0,
                1,
                0
            };
            int dy[] = { 0,
                1,
                0,
                -1
            };

            while (!q.empty())
            {
                int row = q.front().first;
                int col = q.front().second;
                image[row][col] = color;
                q.pop();

                for (int i = 0; i < 4; i++)
                {
                    int newRow = row + dx[i];
                    int newCol = col + dy[i];

                    if (newRow >= 0 && newRow < n && newCol >= 0 && newCol < m && image[newRow][newCol] != color && image[newRow][newCol] == startColor)
                    {
                        q.push({ newRow,
                            newCol });
                    }
                }
            }

            return image;
        }
};

File: 0733-flood-fill/0733-flood-fill.java
class Solution {
    private void dfs(int row, int col, int[][] image, int color, int initialColor, int[] drow, int[] dcol){
        image[row][col] = color;
        int n = image.length;
        int m = image[0].length;
        for(int i = 0; i < 4; i++){
            int nrow = row + drow[i];
            int ncol = col + dcol[i];
            if(nrow >= 0 && nrow < n && ncol >= 0 && ncol < m && image[nrow][ncol] == initialColor && image[nrow][ncol] != color){
                dfs(nrow,ncol,image,color,initialColor,drow,dcol);
            }
        }
    }
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int initialColor = image[sr][sc];
        int[] drow = {1,-1,0,0};
        int[] dcol = {0,0,1,-1};
        dfs(sr,sc,image,color,initialColor,drow,dcol);
        return image;
    }
}